src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+            var members = interfaceType.GetMembers(memberName);
+            if (!members.IsEmpty)
+            {
+                return (IMethodSymbol)members[0];
+            }
+
+            foreach (var current in interfaceType.AllInterfaces)
+            {
+                var member = GetInterfaceMembers(current, memberName);
+                if (member != null)
+                {
+                    return member;
+                }
+            }
+
+            return null;
+        }
+
+        private static bool CheckRequireCollectionStatement(IOperation operation)
+        {
+            return operation.Kind != OperationKind.LocalReference && operation.Kind != OperationKind.FieldReference;
+        }
+
+        private IOperation RemoveImplicitConversion(IOperation collection)
+        {
+            if (collection is IConversionOperation conversion && conversion.IsImplicit)
+            {
+                return RemoveImplicitConversion(conversion.Operand);
+            }
+
+            return collection;
+        }
+
+        private bool CheckForeachVariable(SemanticModel semanticModel, ISymbol foreachVariable, SyntaxNode foreachStatement)
+        {
+            var (start, end) = GetForeachBody(foreachStatement);
+            var dataFlow = semanticModel.AnalyzeDataFlow(start, end);
+
+            if (!dataFlow.Succeeded)
+            {
+                // if we can't get good analysis, assume it is written
+                return false;
+            }
+
+            return !dataFlow.WrittenInside.Contains(foreachVariable);
+        }
+
+        private async Task<Document> ConvertForeachToForAsync(
+            Document document,
+            ForeachInfo foreachInfo,
+            CancellationToken cancellationToken)
+        {
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var workspace = document.Project.Solution.Workspace;
+            var editor = new SyntaxEditor(model.SyntaxTree.GetRoot(cancellationToken), workspace);
+
+            ConvertToForStatement(model, foreachInfo, editor, cancellationToken);
+
+            var newRoot = editor.GetChangedRoot();
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        protected class ForeachInfo
+        {
+            public ForeachInfo(
+                string countName, string explicitCastInterface, bool requireCollectionStatement, SyntaxNode foreachStatement)
+            {
+                CountName = countName;
+
+                ExplicitCastInterface = explicitCastInterface;
+                RequireExplicitCast = explicitCastInterface != null;
+                RequireCollectionStatement = requireCollectionStatement || RequireExplicitCast;
+
+                ForeachStatement = foreachStatement;
+            }
+
+            public string CountName { get; }
+
+            public bool RequireExplicitCast { get; }
+            public string ExplicitCastInterface { get; internal set; }
```

Comment:
why is this internal-set?

Reply:
auto property generator thing.

Reply:
?  i don't understand what that means.  it doesn't look like you ever write to this outside of hte constructor.  so i don't see why it's internal-set.

Reply:
I meant, I used refactoring to generate the property. and didn't clean it up.